payjoin/src/uri.rs

```diff
@@ -1,101 +1,165 @@
-            Payjoin::Unsupported => false,
-        }
+impl<'a, N: NetworkValidation, P: DeserializeParams<'a> + DeserializationError + Sized>
+    From<bip21::Uri<'a, N, P>> for PayjoinUri<'a, N, P>
+{
+    fn from(value: bip21::Uri<'a, N, P>) -> Self { Self { inner: value } }
+}
+
+impl<'a> PayjoinUri<'a, NetworkUnchecked, Payjoin> {
+    /// Marks network of this address as checked.
+    pub fn assume_checked(self) -> PayjoinUri<'a, NetworkChecked, Payjoin> {
+        PayjoinUri::new(
+            self.inner.address.assume_checked(),
+            self.inner.extras,
+            self.inner.amount,
+            self.inner.label,
+            self.inner.message,
+        )
+        .into()
+    }
+    /// Checks whether network of this address is as required.
+    pub fn require_network(
+        self,
+        network: Network,
+    ) -> Result<PayjoinUri<'a, NetworkChecked, Payjoin>, Error> {
+        Ok(PayjoinUri::new(
+            self.inner.address.require_network(network)?,
+            self.inner.extras,
+            self.inner.amount,
+            self.inner.label,
+            self.inner.message,
+        )
+        .into())
     }
 }
 
-pub struct PayjoinParams {
-    pub(crate) _endpoint: Url,
-    pub(crate) disable_output_substitution: bool,
+impl<'a, N: NetworkValidation> PayjoinUri<'a, N, Payjoin> {
+    fn new(
+        address: Address<N>,
+        extras: Payjoin,
+        amount: Option<Amount>,
+        label: Option<bip21::Param<'a>>,
+        message: Option<bip21::Param<'a>>,
+    ) -> Self {
+        let mut uri = bip21::Uri::with_extras(address, extras);
+        uri.amount = amount;
+        uri.label = label;
+        uri.message = message;
+        Self { inner: uri }
+    }
+    /// Returns the receiver bitcoin address from this uri.
+    pub fn address(&self) -> &Address<N> { &self.inner.address }
+    /// Returns the amount in bitcoin from this uri.
+    pub fn amount(&self) -> Option<Amount> { self.inner.amount }
+    /// Returns the endpoint url from this uri.
+    pub fn endpoint(&self) -> &Url { self.inner.extras.endpoint() }
+    /// Returns whether output substitution is disabled for this uri.
+    pub fn disable_output_substitution(&self) -> bool {
+        self.inner.extras.disable_output_substitution()
+    }
+    /// Returns the label from this uri.
+    pub fn label(&self) -> Option<String> {
+        match self.inner.label.clone() {
+            None => None,
+            Some(label) => {
+                let label: Cow<'_, str> = label.try_into().unwrap();
+                Some(label.to_string())
+            }
+        }
+    }
+    /// Returns the message from this uri.
+    pub fn message(&self) -> Option<String> {
+        match self.inner.message.clone() {
+            None => None,
+            Some(message) => {
+                let message: Cow<'_, str> = message.try_into().unwrap();
+                Some(message.to_string())
+            }
+        }
+    }
```

Comment:
could we eliminate this complexity and unhandled error by exposing `inner` or just making an `inner` alias the public interface as it was before?

Reply:
removed the getters from `PayjoinUri` and changed `inner` to `pub`